### ISSUE TITLE
Changing a FormatString property for a Date column

### ISSUE THREAD

**Issue by SergiyVV** [otykier/TabularEditor#684]
Hi,

There is a pbix file. The data model consists of tables that are in Import mode. I need to change a format string of all the columns that are of Date type. The format of these Date type columns is not custom. It is a predefined format - "General Date"
My aim is to change the format from "General Date" to "d/m/yyyy" (for example).

I fail to achieve my aim via Tabular Editor.
I tried two approaches.
1.
I made changes via Tabular Editor UI. I changed "Format String" property of a target column using "d/m/yyyy" as a new value, then saved the changes to be reflected in pbix file. It seemed everything went OK, but in reality I didn't succeed. This column in Power BI desktop still had the same old "General Date" format.
2.
I executed the following code:
`foreach(var column in Model.Tables.SelectMany(t => t.Columns)){
    if(column.DataType.ToString() == "DateTime") {
        column.FormatString = "d/m/yyyy";
    }
}`
The result was the same as before. The columns in Power BI desktop still had the same old "General Date" format.

What I noticed is if a column I was working on had not a predefined format but a custom one, then each time I was successful. To make it custom I switched back to Power BI desktop and in a Properties pane for a target column in a "Date time format" drop down I selected "Custom". Once I did that, It allowed me to use Tabular Editor to change a FormatString property at any time and see the result reflected in Power BI desktop.

So, is there a way to somehow change a FormatString property for a Date column with "General Date" format and see this change reflected in Power BI desktop? What should be done?

Thanks in advance!

**Comment by otykier** [comment]
The problem is that Power BI is maintaining an annotation called "Format" which seems to override the FormatString property. Other users have managed to get it working by explicitly deleting the Format annotation. So in your case, you could extend your script a little bit like so:

```csharp
foreach(var column in Model.Tables.SelectMany(t => t.Columns))
{
    if(column.DataType.ToString() == "DateTime")
    {
        column.FormatString = "d/m/yyyy";
        column.RemoveAnnotation("Format");
    } 
}
```

Please let me know if this helps, thanks!

**Comment by SergiyVV** [comment]
Thank you for answering,
I didn't succeed, though.

![NothingHappened](https://user-images.githubusercontent.com/52447998/99580556-41525b80-29e8-11eb-9b86-fe6a8548db26.png)

As you see on the picture the format string hasn't changed.

I knew nothing of Annotations, so I decided to check the count and values of Annotations before and after executing of RemoveAnnotation("Format"). It happens that before and after the execution of RemoveAnnotation("Format") I get the same results, namely:
Model.Tables["Table1"].Columns["Date"].GetAnnotationsCount().Output();     
returns 6
Model.Tables["Table1"].Columns["Date"].GetAnnotation(2).Output();              
returns the following string:

```
<Format Format="DateTimeGeneralPattern">
	<DateTimes>
		<DateTime LCID="1033" Group="ShortDate" FormatString="M/d/yyyy" />
	</DateTimes>
</Format>
```

But if I switch back to Power BI desktop and set "Date time format" as "Custom", then return to Tabular Editor and check 

``
Model.Tables["Table1"].Columns["Date"].GetAnnotation(2).Output();  
``

then I get the string that I passed before with a format I need:

``
<Format Format="Custom" Custom="d/M/yyyy" />
``

Now I am able to set any format string I need using Tabular Editor.

But the problem still remains - I need to get rid of a predefined date format. 
Is it possible to solve?

**Comment by otykier** [comment]
In general, I believe auto date/time causes a number of issues when an external tool makes changes to a Power BI model. The recommendation is to always turn auto date/time off, and provide your own calendar table. Could you try to see if this perhaps makes any difference to the formatting issue?

I'm not sure if there is a better way, but in general you can recognise Power BI-created date tables from their name: LocalDateTable_<GUID>. That name pattern should be easily detectable with a script.

**Comment by SergiyVV** [comment]
Thank you, Daniel.

"Auto date/time" setting does not make any difference. In my production pbix file auto creation of date tables is turned off. 
So, the only way to reach my goal, as I see it now, is to make changes in Power BI manually changing predefined date format to a custom one(no matter what format it will be). Having done that I'll be able to use Tabular Editor each time when there is a need to change date format that meets customer requirements.

Should I close the issue now?

**Comment by otykier** [comment]
I just took a closer look. I must admit that I haven't spend a lot of time looking into how formatting behaves on columns (been focusing mostly on measures, where it works great if you remove the Format annotation manually).

Here are my findings: If I use a date column in a slicer in List or Dropdown mode, I was able to get the dates to format correctly. Same for a Matrix visual. On the report view and data view pages, the format I assigned in Tabular Editor ("d/M/yy") also shows up correctly in the ribbon:

![image](https://user-images.githubusercontent.com/8976200/100059225-0064a780-2e2b-11eb-9d33-287c1bd38247.png)

However, when using a slicer in "Between" mode, it doesn't respect the formatting, which can also be seen on the screenshot above. And on the model view page, it seems that the format dropdown completely ignores the setting within Tabular Editor, as you also noticed.

These latter issues are definitely bugs in Power BI Desktop, since the behaviour is inconsistent. I recommend you post this issue to the Power BI forums - I will reach out to the team internally to also make them aware of this issue.
